hack/sync-forks.sh

```diff
@@ -0,0 +1,106 @@
+    mkdir -p "${toRepoLocation}"
+
+    pushd "${fromRepoLocation}" >/dev/null
+    # if we already have a git repo, assume it's the right one and get the requested branch
+    git rev-parse --is-inside-work-tree || git clone --single-branch --branch=${fromBranch} -- "${fromRepo}" ${fromRepoLocation}
+    git fetch --refmap -- origin refs/heads/${fromBranch}:refs/remotes/origin/${fromBranch}
+    git checkout ${fromBranch}
+    git reset --hard origin/${fromBranch}
+    firstFromSHA=$(git log --reverse --oneline  --format='%H' | head -n 1)
+    newFromSHA=$(git log --oneline --format='%H' -1)
+    popd >/dev/null
+
+    pushd "${toRepoLocation}" >/dev/null
+    # if we aren't in a git repo, clone it
+    git rev-parse --is-inside-work-tree || git clone --single-branch --branch=${toBranch} -- "${toRepo}" ${toRepoLocation}
+    git fetch --refmap -- origin refs/heads/${toBranch}:refs/remotes/origin/${toBranch}
+    git checkout ${toBranch}
+    git reset --hard origin/${toBranch}
+    startingFromSHA=${firstFromSHA}
+    if [ -f "${fromRepoName}.sha" ]; then
+        startingFromSHA=$(cat ${fromRepoName}.sha)
+    fi
+    if [ "${newFromSHA}" == "${startingFromSHA}" ]; then
+        popd >/dev/null
+        echo "already at level: ${newFromSHA}, nothing to sync"
+        exit 0
+    fi
+    echo "syncing from ${startingFromSHA}..${newFromSHA}"
+    # create a clean branch to start from
+    git branch -D "${toBranch}-working" || true
+    git checkout -b "${toBranch}-working"
+    popd >/dev/null
+
+    pushd "${fromRepoLocation}" >/dev/null
+    patchDir="${TMPDIR}/patches"
+    rm -rf ${patchDir} && mkdir -p "${patchDir}"
+    index=0
+    for commit in $(git log --format='%H' --no-merges --reverse -- "${startingFromSHA}..${newFromSHA}" "${fromDir}"); do
+      git format-patch --raw --start-number=${index} --relative="${fromDir}" "${commit}^..${commit}" -o "${patchDir}"
+      let index+=10
+    done
+    # remove all commits that had no entries
+    find "${patchDir}" -type f -size 0 -exec rm {} \;
+    popd >/dev/null
+
+    # only patch if we have patches
+    if [ "$(ls -A ${patchDir})" ]; then
+        # apply the changes
+        if ! git am -3 --ignore-whitespace ${patchDir}/*.patch; then
+          echo 2>&1
+          echo "++ Patches do not apply cleanly, continue with 'git am' in ${relativedir}" 2>&1
+          exit 1
+        fi
+        echo 2>&1
+        echo "++ All patches applied cleanly upstream" 2>&1
+    else
+        echo "No patches to apply.  Bumping sync commit."
+    fi
+
+    echo "${newFromSHA}" > ${fromRepoName}.sha
+    git add ${fromRepoName}.sha && git commit -m "sync(${fromRepo}): ${newFromSHA}"
+
+    # push the results
+    echo "Don't forget to push your changes with:"
+    echo "git push origin ${toBranch}-working:${toBranch}"
+}
+
+echo "++ Temporary directory is: ${TMPDIR}"
+
+sync_fork "github.com/openshift/origin" "master" "vendor/k8s.io/kubernetes" "github.com/openshift/kubernetes" "release-1.7.0"
```

Comment:
And a third script that builds a script with all these lines based on some metadata file with a verify script that makes sure we have a line for every dep we carry upstreams for.

I could also go for a script that drives the `sync_fork` based on the metadata file.  That probably makes more sense.

Reply:
+100 to machine-readable config